### ISSUE TITLE
False positive E0308 - mismatched types in TrustDNS code

### ISSUE THREAD

**Issue by msehnout** [intellij-rust/intellij-rust#2542]
Source code: https://github.com/bluejekyll/trust-dns.git
False positive in file `src/authority/catalog.rs` line `313` (see the picture)
`query ` is of type LowerQuery and the `name` function returns `LowerName`: https://github.com/bluejekyll/trust-dns/blob/master/client/src/op/lower_query.rs#L38

The code compiles just fine, no errors are produced.
```
$ rustc --version 
rustc 1.26.0 (a77568041 2018-05-07)
```
Plugin version: `Version: 0.2.0.2094-181`

![rust-ide-bug](https://user-images.githubusercontent.com/9369632/40272933-0bcdac00-5bb7-11e8-943c-886923e75bb4.png)

**Comment by Undin** [comment]
Hm, I don't get this error but I get other type checker error in the same file
<img width="804" alt="screen shot 2018-05-21 at 12 02 19" src="https://user-images.githubusercontent.com/2539310/40299415-da4e4426-5cee-11e8-9ae7-62fd42ebae63.png">
@vlad20012 Looks like that type inference didn't find `Borrow` impl for `LowerName` struct
